contributors/design-proposals/selector-generation-from-kubectl.md

```diff
@@ -0,0 +1,174 @@
+Selector generation from kubectl
+==================
+
+# Goals
+
+Make `kubectl apply` work well with selector.
+
+# Problem Description
+
+The field `spec.selector` of Deployments, ReplicaSet, DaemonSet and ReplicationController are defaulted from `spec.template.metadata.labels`, if it is unspecified.
+And there is validation to assure `spec.selector` always selects the pod template.
+The defaulting of selector may prevent from `kubectl apply` from working when updating
+the `spec.selector` field. e.g. [kubernetes/kubernetes#26202 (comment)](https://github.com/kubernetes/kubernetes/issues/26202#issuecomment-221421254)
+
+# Proposed changes
+
+We can let `kubectl` do something similar to [Selector Generation](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/selector-generation.md) on creation. We can let `kubectl` add a set of labels to assure non-overlapping. 
+There are no changes on server-side and the server-side defaulting is still kept for backward compatible reason.
+
+## API
+
+No required change.
+
+## kubectl
+
+Add client-side selector generation, which is disabled by default and can be enabled by flag in `kubectl`.
+
+`kubectl` will add a set of labels to `spec.selector` and `spec.template.metadata.labels` before sending the object to the API server. 
+
+The generated labels will not be included in the `kubectl.kubernetes.io/last-applied-configuration` annotation, if it is created by `kubectl create --save-config` or `kubectl apply`.
+
+### Generated labels
+
+There are at least 3 sets of labels to choose:
+
+The first set of labels is:
+
+- Add 3 more labels to `spec.selector`
+  - `matchLabels["resource-group"]="$GROUP"`
+  - `matchLabels["resource-kind"]="$KIND"`
+  - `matchLabels["resource-name"]="$NAME"`
+- Append 3 labels to `spec.template.metadata.labels`
+  - `resource-group=$GROUP`
+  - `resource-kind=$KIND`
+  - `resource-name=$NAME`
+
+The second set of labels is:
+
+- Add 1 more label, which concatenates all the three in the first set, to `spec.selector`
+  - `matchLabels["controllers.k8s.io/selector"]="$GROUP/$KIND/$NAME"`
```

Comment:
Why is this needed if we have the other 3 labels?

Also, I try to avoid label keys with prefixes.